### ISSUE TITLE
False positive `reportGeneralTypeIssue` reported when applying decorators that make use of Protocols on class methods

### ISSUE THREAD

**Issue by zoopp** [microsoft/pyright#3253]
**Describe the bug**
A false positive `reportGeneralTypeIssue`  error is reported when decorating class methods with decorators that make use of protocols.

**To Reproduce**
Invoke pyright on the following code:

```python
import functools
from typing import Protocol


class Handler(Protocol):
    def __call__(self, value: int) -> int:
        ...


def decorator(handler: Handler) -> Handler:
    @functools.wraps(handler)
    def wrapper(*args, **kwargs):
        return handler(*args, **kwargs)

    return wrapper


@decorator
def handle_foo(value: int) -> int:				 # <- This is fine
    return value + 1


class FooHandler:
    @decorator                                            # <- Error reported here
    def handle_foo(self, value: int) -> int:
        return value + 1

    bar = decorator(handle_foo)					# <- This is also fine
```

**Expected behavior**
Error should not be reported because the method passed to the decorator does implement the Handler protocol.

**Screenshots or Code**
```
No configuration file found.
No pyproject.toml file found.
stubPath /home/zoopp/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
/home/zoopp/test.py
  /home/zoopp/test.py:24:6 - error: Argument of type "(self: Self@FooHandler, value: int) -> int" cannot be assigned to parameter "handler" of type "Handler" in function "decorator"
    Type "(self: Self@FooHandler, value: int) -> int" cannot be assigned to type "(value: int) -> int"
      Parameter name mismatch: "value" versus "self"
      Parameter 1: type "int" cannot be assigned to type "Self@FooHandler"
        "int" is incompatible with "FooHandler"
      Keyword parameter "value" is missing in destination (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
Completed in 1.034sec
```

**VS Code extension or command-line**
I'm using pyright 1.1.232 from the command line.

**Comment by erictraut** [comment]
This isn't a false positive. The decorator cannot be applied because of a type violation. Mypy emits the same error in this case.

The problem is that the decorator is defined to accept a callback function that takes only one parameter — an `int` value. You're trying to apply the decorator to a method that takes two parameters — including a `self`. If you want your decorator to work for both methods and non-methods, you'll need to annotate it differently.
